apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/data_updater/impl.ex

```diff
@@ -71,11 +75,123 @@ defmodule Astarte.DataUpdaterPlant.DataUpdater.Impl do
     stats_and_introspection =
       Queries.retrieve_device_stats_and_introspection!(new_state.realm, device_id)
 
+    updated_state = Map.merge(new_state, stats_and_introspection)
+
+    add_device_to_rpc_handler_state(encoded_device_id, updated_state.realm, updated_state.groups)
+
     # TODO this could be a bang!
     {:ok, ttl} = Queries.get_datastream_maximum_storage_retention(new_state.realm)
 
-    Map.merge(new_state, stats_and_introspection)
-    |> Map.put(:datastream_maximum_storage_retention, ttl)
+    Map.put(updated_state, :datastream_maximum_storage_retention, ttl)
+  end
+
+  def add_device_to_rpc_handler_state(device_id, realm, groups) do
+    device = %Astarte.DataUpdaterPlant.RPC.Device{
+      device_id: device_id,
+      realm: realm,
+      groups: groups
+    }
+
+    Astarte.DataUpdaterPlant.RPC.DataUpdater.add_device(device)
+    Logger.info("Device added to dup_rpc_handler state: #{inspect(device_id)}")
+  end
+
+  def remove_device_from_rpc_handler_state(device_id) do
+    Astarte.DataUpdaterPlant.RPC.DataUpdater.remove_device(device_id)
+    Logger.info("Device removed from dup_rpc_handler state: #{inspect(device_id)}")
+  end
+
+  def handle_install_persistent_trigger(
+        %State{connected: false} = state,
+        _object_id,
+        _object_type,
+        _parent_id,
+        _trigger_id,
+        _trigger,
+        _target
+      ) do
+    Logger.info("Device is not connected. Skipping persistent trigger installation.")
+    {:ok, state}
+  end
```

Comment:
This seems an interesting optimization.
Why would you not install the trigger for disconnected devices?
I reckon that a device could connect right after installing a trigger, and it would still be useful to have DataUpdater honor the trigger (without waiting for N minutes before DataUpdater polls existing triggers from the database).

Let's evaluate this point. If we end up noticing that the distinction of connected/disconnected devices is not needed, we could drastically simplify the pull request since no queries or live state would be needed to keep track of connected devices: we could just notify the existing DataUpdater processes.